{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Set up notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-11T18:48:36.491695Z",
     "start_time": "2020-03-11T18:48:36.481458Z"
    },
    "code_folding": [
     0
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Import packages\n",
    "%matplotlib inline\n",
    "import os\n",
    "import re\n",
    "import string\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-11T18:48:17.597037Z",
     "start_time": "2020-03-11T18:48:17.553539Z"
    },
    "code_folding": [
     0,
     1
    ],
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Define functions (from Kimiko)\n",
    "def get_ref_and_align(filename):\n",
    "    na = string.ascii_uppercase+'+ '\n",
    "    with open(filename) as f:\n",
    "        seqs = []\n",
    "        alignments = []\n",
    "        new_ref = ''\n",
    "        new_align = ''\n",
    "        for i, line in enumerate(f):\n",
    "            if not line:\n",
    "                pass\n",
    "            elif line[0] == '>':\n",
    "                if new_ref:\n",
    "                    seqs.append(new_ref)\n",
    "                    new_ref == ''\n",
    "                if new_align:\n",
    "                    alignments.append(new_align)\n",
    "#                     alignments.append(new_align)\n",
    "                    new_ref = ''\n",
    "                    new_align = ''\n",
    "            elif line[0:6] == 'Query ' and not seqs:\n",
    "                new_ref += line.strip().split()[2]\n",
    "            elif line.count(' ') >= 12 and all(i in na for i in line.strip()):\n",
    "                new_align += line.rstrip('\\n').lstrip(' ')\n",
    "        alignments.append(new_align)\n",
    "        seqs.append(alignments)\n",
    "        return seqs  \n",
    "    \n",
    "def find_motif_site(motif, MaCS):#array of HOG components sequences from reference genome\n",
    "    # Finds all of the Hog1 concensus sites in the ref\n",
    "    # str.find() only returns the first instance of the site\n",
    "    return [m.start() for m in re.finditer(motif,MaCS)]\n",
    "\n",
    "def number_concensus_sites(seqs):\n",
    "    sp, tp = find_concensus_site(seqs[0])\n",
    "    ref_num = len(sp+tp)\n",
    "    align_num = []\n",
    "    for seq in seqs[1]:\n",
    "        sp, tp = find_concensus_site(seq)\n",
    "        num = len(sp+tp)\n",
    "        align_num.append(num)\n",
    "    MaCS_hist = plt.hist(align_num)\n",
    "#     MaCS_hist[0].set_color('r')\n",
    "#     plt.hist().get_xticklabels()[0].set_color('red') \n",
    "    plt.show()\n",
    "#     return [ref_num,align_num]\n",
    "\n",
    "# def plot_concensus_sites(seqs):\n",
    "#     sp, tp = find_concensus_site(seqs[0])\n",
    "#     plt.bar(sp,[1,1,1],color='g')\n",
    "#     plt.show()\n",
    "\n",
    "def compare_concensus_sites(seqs,filename):#array of HOG components sequences from reference genome\n",
    "    # Finds all of the Hog1 concensus sites in the ref\n",
    "    # str.find() only returns the first instance of the site\n",
    "#     for x in seqs:\n",
    "#         print(x)\n",
    "#         _ref = [m.start() for m in re.finditer('SP', x)]\n",
    "#         print(_ref)\n",
    "    ref_sp, ref_tp = find_concensus_site(seqs[0])\n",
    "    if ref_sp or ref_tp:\n",
    "#     ref['ID'] = 'ref'\n",
    "        sps = []\n",
    "        tps = []\n",
    "        for seq in seqs[1]:\n",
    "            seq_sp, seq_tp = find_concensus_site(seq)\n",
    "    #         seq['ID'] = str(i)\n",
    "            sps += seq_sp\n",
    "            tps += seq_tp\n",
    "    #     print(sps)\n",
    "        plt.clf()\n",
    "        plt.rcParams['figure.figsize'] = [20.0, 8.0]\n",
    "        if sps:\n",
    "            sp_labels, sp_count = zip(*Counter(sps).items())\n",
    "            plt.bar(sp_labels, sp_count, width=1, color = 'g', label='SP')\n",
    "        if tps:\n",
    "            tp_labels, tp_count = zip(*Counter(tps).items())\n",
    "            plt.bar(tp_labels, tp_count, width=1, color='b', label='TP')\n",
    "    #     print(sp_count)\n",
    "    #     indices = np.arange(len(labels))\n",
    "        plt.axhline(y=len(seqs[1]),color='r', linestyle='--', label='Total number of sequences analyzed')\n",
    "        plt.legend(bbox_to_anchor=[1, 0.5], loc='center left')\n",
    "        plt.xlim(0,len(seqs[0]))\n",
    "        plt.xlabel('Amino Acid Sequence Position')\n",
    "        plt.title(filename[0].upper() + filename[1:4])\n",
    "        plt.ylabel('Number of sequences')\n",
    "        plt.show()\n",
    "    else:\n",
    "        print('Bug in code')\n",
    "#         ref = ref.append(seq, ignore_index=True)\n",
    "#     _aligns = [[m.start() for m in re.finditer('SP', seqs[1])], [m.start() for m in re.finditer('TP', seqs[1])]]\n",
    "#     return [m.start() for m in re.finditer('SP',seqs)], [m.start() for m in re.finditer('TP',seqs)] \n",
    "#     print(ref)\n",
    "#     sns.barplot(x='locations', y)\n",
    "#     return ref, seq\n",
    "\n",
    "def plot_seq_MCS(seqs,filename):\n",
    "    seqs = [seqs[0]] + seqs[1]\n",
    "    grid =[]\n",
    "    for x in seqs:\n",
    "        temp = []\n",
    "        for i in range(len(x)-1):\n",
    "        #     print(x[i:i+2])\n",
    "            if x[i:i+2] == 'SP':\n",
    "                temp.append(2)\n",
    "            elif x[i:i+2] == 'TP':\n",
    "                temp.append(1)\n",
    "            else:\n",
    "                temp.append(0)\n",
    "        grid.append(temp)\n",
    "    colors = ['#DCDCDC','#ee9537','#3db7d6']\n",
    "    grid_df = pd.DataFrame(grid)\n",
    "    \n",
    "    plt.clf()\n",
    "    plt.rcParams['figure.figsize'] = [30.0, 20.0]\n",
    "    sns.heatmap(grid_df, cmap=colors, cbar=False)\n",
    "    plt.title(filename[0].upper() + filename[1:4]) #add percent conserved\n",
    "    plt.xlabel('Amino Acid Sequence Position')\n",
    "    plt.ylabel('Sequence number') # change to strain name!\n",
    "    plt.show()\n",
    "    \n",
    "def get_MaCS_df(seq):\n",
    "    refs_sp = [m.start() for m in re.finditer('SP', seq)]\n",
    "    refs_tp = [m.start() for m in re.finditer('TP', seq)]\n",
    "    _refs_locs = refs_sp + refs_tp\n",
    "    _ = {'locations': _refs_locs, 'type': ['SP' for x in range(len(refs_sp))] + ['TP' for x in range(len(refs_tp))]}\n",
    "    df = pd.DataFrame.from_dict(_)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make pandas data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-11T20:17:29.831388Z",
     "start_time": "2020-03-11T20:17:28.247439Z"
    },
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#### USER DEFINED THINGS ####\n",
    "# Directory with protein sequences\n",
    "dir_to_check = '/Users/AmyAllen/Documents/A-pipeline-for-fast-conserved-protein-motif-identification/aligned_cp_for_counting/trimmed_data/'\n",
    "# Motif sequence \n",
    "motif = 'TP'\n",
    "#### END USER DEFINED THINGS #### \n",
    "\n",
    "# Set all empty lists to be filled in for loops\n",
    "protein_list = []\n",
    "short_protein_list = []\n",
    "num = []\n",
    "motif_loc = []\n",
    "all_locations = []\n",
    "motif_list = []\n",
    "n_loc = []\n",
    "conserved = []\n",
    "\n",
    "# Get all files from direcotry \n",
    "files = os.listdir(dir_to_check)\n",
    "\n",
    "# For loop to loop through all files in directory \n",
    "for filename in files:\n",
    "    seq = get_ref_and_align(dir_to_check+filename) # Import protein sequence from filename\n",
    "    protein = filename.split('.')[0].split('_')[0] # Get name of protein (string before the _ or .)\n",
    "    \n",
    "    # For loop to loop through each instance of the protein \n",
    "    for i in range(len(seq[1])):\n",
    "        motif_loc+=(find_motif_site(motif, seq[1][i])) # Find motif sites in protein  \n",
    "        \n",
    "    locations = list(dict.fromkeys(motif_loc)) # Store motif sites \n",
    "    n_loc += [len(locations)] # make list of number of locations in filename \n",
    "    all_locations+=locations # make list of all locations \n",
    "    \n",
    "    protein_num = []\n",
    "    # For loop to loop through all locations \n",
    "    for loc in locations:\n",
    "        num.append(motif_loc.count(loc)) # Count number of instances (all proteins)\n",
    "        protein_num.append(motif_loc.count(loc)) # Count number of instances (protein associated with file name)\n",
    "        \n",
    "    conserved.append(np.mean(protein_num)/len(seq[1])*100) # Calculate how conserved the sites in the protein are between instances \n",
    "    protein_list+=[protein] * len(locations) # List of proteins for detailed\n",
    "    short_protein_list+=[protein] # List of proteins for short\n",
    "    motif_list+=[motif] * len(locations) # list of motifs for deailed \n",
    "\n",
    "# Make dictionaries of key information for the detailed and summary data tables\n",
    "detail_data = {'Protein': protein_list, 'Motif': motif_list, 'Location' : all_locations, 'Frequency' : num}\n",
    "summary_data = {'Protein' : short_protein_list, 'Number of locations' : n_loc, 'Conservedness' : conserved}\n",
    "\n",
    "# Make data frames \n",
    "detailed_data_frame = pd.DataFrame.from_dict(detail_data)\n",
    "summary_data_frame = pd.DataFrame.from_dict(summary_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-11T20:02:37.693067Z",
     "start_time": "2020-03-11T20:02:37.643577Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Motif</th>\n",
       "      <th>Location</th>\n",
       "      <th>Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ste50</td>\n",
       "      <td>TP</td>\n",
       "      <td>243</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ste50</td>\n",
       "      <td>TP</td>\n",
       "      <td>340</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ste50</td>\n",
       "      <td>TP</td>\n",
       "      <td>242</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ste50</td>\n",
       "      <td>TP</td>\n",
       "      <td>339</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hot1</td>\n",
       "      <td>TP</td>\n",
       "      <td>243</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein Motif  Location  Frequency\n",
       "0   ste50    TP       243         90\n",
       "1   ste50    TP       340         90\n",
       "2   ste50    TP       242          3\n",
       "3   ste50    TP       339          3\n",
       "4    hot1    TP       243         90"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check detailed data frame\n",
    "detailed_data_frame.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2020-03-11T20:02:42.243769Z",
     "start_time": "2020-03-11T20:02:42.230828Z"
    },
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein</th>\n",
       "      <th>Number of locations</th>\n",
       "      <th>Conservedness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cdc42</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ste50</td>\n",
       "      <td>4</td>\n",
       "      <td>50.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hot1</td>\n",
       "      <td>4</td>\n",
       "      <td>50.543478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sho1</td>\n",
       "      <td>6</td>\n",
       "      <td>66.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sln1</td>\n",
       "      <td>15</td>\n",
       "      <td>66.449275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Protein  Number of locations  Conservedness\n",
       "0   cdc42                    0            NaN\n",
       "1   ste50                    4      50.000000\n",
       "2    hot1                    4      50.543478\n",
       "3    sho1                    6      66.666667\n",
       "4    sln1                   15      66.449275"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check summary data frame \n",
    "summary_data_frame.head()"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [python3]",
   "language": "python",
   "name": "Python [python3]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "48px",
    "width": "251px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
